dataclients/kubernetes/logger.go

```diff
@@ -1,32 +1,62 @@
 package kubernetes
 
-import log "github.com/sirupsen/logrus"
+import (
+	"fmt"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+)
 
 type logger struct {
 	logger *log.Entry
+
+	mu      sync.Mutex
+	history map[string]struct{}
 }
 
+// newLogger creates a logger that logs each unique message once
+// for the resource identified by kind, namespace and name.
+// It logs nothing when disabled.
 func newLogger(kind, namespace, name string, enabled bool) *logger {
 	if !enabled {
 		return nil
 	}
-	return &logger{log.WithFields(log.Fields{"kind": kind, "ns": namespace, "name": name})}
+	return &logger{logger: log.WithFields(log.Fields{"kind": kind, "ns": namespace, "name": name})}
 }
 
 func (l *logger) Debugf(format string, args ...any) {
 	if l != nil {
-		l.logger.Debugf(format, args...)
+		l.once(log.DebugLevel, format, args...)
 	}
 }
 
 func (l *logger) Infof(format string, args ...any) {
 	if l != nil {
-		l.logger.Infof(format, args...)
+		l.once(log.InfoLevel, format, args...)
 	}
 }
 
 func (l *logger) Errorf(format string, args ...any) {
 	if l != nil {
-		l.logger.Errorf(format, args...)
+		l.once(log.ErrorLevel, format, args...)
+	}
+}
+
+func (l *logger) once(level log.Level, format string, args ...any) {
+	if !l.logger.Logger.IsLevelEnabled(level) {
+		return
+	}
+	l.mu.Lock()
```

Comment:
Is it okay to create high lock contention on the logger (from the quick look it seems it must be used very frequently)?

Reply:
Mustafa pointed out we have some loggingInterval constant https://github.com/zalando/skipper/blob/v0.18.14/dataclients/kubernetes/logger_test.go#L42 which defines how often logger will produce output.
Let's imagine l.Errorf is callel 1_000_000 times in the single loggingInterval. In this case, will this mutex be locked 1_000_000 times? Or only once?

Reply:
Dataclient is actually single threaded so there is no lock contention, I added mutex because underlying logger is thread-safe so I wanted to keep this property.

We ratelimit logs to log every minute, see https://github.com/zalando/skipper/pull/2427

This change eliminates duplicate messages within single update cycle.